policybot/cmd/server.go

```diff
@@ -115,3 +151,242 @@ func serverCmd() *cobra.Command {
 type dummyIoWriter struct{}
 
 func (dummyIoWriter) Write([]byte) (int, error) { return 0, nil }
+
+// Server represents a running bot instance.
+type server struct {
+	listener   net.Listener
+	shutdown   sync.WaitGroup
+	httpServer http.Server
+
+	clientID     string
+	clientSecret string
+	secretState  string
+
+	dashboard *dashboard.Dashboard
+}
+
+// Runs the server.
+//
+// If config comes from a container-based file, this will try to run the server, but if
+// problems occur (probably due to bad config), then the function returns with an error.
+//
+// If config comes from a repo-based file, this will also try to run the server, but if an error
+// occurs, it will refetch the config every minute and try again. And so in that case, this
+// function never returns.
+func runServer(base *config.Args) error {
+	for {
+		// copy the baseline config
+		c := *base
+		cfg := &c
+
+		// load the config file
+		if err := cfg.Fetch(); err != nil {
+			if cfg.StartupOptions.ConfigRepo != "" {
+				log.Errorf("Unable to load configuration file, waiting for 1 minute and then will try again: %v", err)
+				time.Sleep(time.Minute)
+				continue
+			} else {
+				return fmt.Errorf("unable to load configuration file: %v", err)
+			}
+		}
+
+		if err := runWithConfig(cfg); err != nil {
+			if cfg.StartupOptions.ConfigRepo != "" {
+				log.Errorf("Unable to initialize server likely due to bad config, waiting for 1 minute and then will try again: %v", err)
+				time.Sleep(time.Minute)
+			} else {
+				return fmt.Errorf("unable to initialize server: %v", err)
+			}
+		} else {
+			log.Infof("Configuration change detected, attempting to reload configuration")
+		}
+	}
+}
+
+func httpsRedirectHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		proto := req.Header.Get("x-forwarded-proto")
```

Comment:
Comment about what this is? (I assume x-forwarded-proto is something that you use internally?

Reply:
This was introduced by Mandar in a previous PR, I just moved it around.

@mandarjog